Content.Shared/Anomaly/Effects/BluespaceAnomalySystem.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System.Linq;
+using Content.Shared.Anomaly.Components;
+using Content.Shared.Anomaly.Effects.Components;
+using Content.Shared.Mobs.Components;
+using Content.Shared.Teleportation.Components;
+using Robust.Shared.Map;
+using Robust.Shared.Map.Components;
+using Robust.Shared.Random;
+namespace Content.Shared.Anomaly.Effects;
+
+public sealed class BluespaceAnomalySystem : EntitySystem
+{
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly EntityLookupSystem _lookup = default!;
+    [Dependency] private readonly SharedTransformSystem _xform = default!;
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<BluespaceAnomalyComponent, AnomalyPulseEvent>(OnPulse);
+        SubscribeLocalEvent<BluespaceAnomalyComponent, AnomalySupercriticalEvent>(OnSupercritical);
+        SubscribeLocalEvent<BluespaceAnomalyComponent, AnomalySeverityChangedEvent>(OnSeverityChanged);
+    }
+
+    private void OnPulse(EntityUid uid, BluespaceAnomalyComponent component, ref AnomalyPulseEvent args)
+    {
+        var xform = Transform(uid);
+        var range = component.MaxShuffleRadius * args.Severity;
+        var allEnts = _lookup.GetComponentsInRange<MobStateComponent>(xform.Coordinates, range)
+            .Select(x => x.Owner).ToList();
+        allEnts.Add(uid);
+
+        var xformQuery = GetEntityQuery<TransformComponent>();
+        var coords = new List<EntityCoordinates>();
+        foreach (var ent in allEnts)
+        {
+            if (xformQuery.TryGetComponent(ent, out var xf))
+                coords.Add(xf.Coordinates);
+        }
+
+        while (allEnts.Any())
+        {
+            var ent = _random.PickAndTake(allEnts);
+            var coord = _random.PickAndTake(coords);
+
+            _xform.SetCoordinates(ent, coord);
+            _audio.PlayPvs(component.TeleportSound, ent);
+        }
+    }
+
+    private void OnSupercritical(EntityUid uid, BluespaceAnomalyComponent component, ref AnomalySupercriticalEvent args)
+    {
+        var xform = Transform(uid);
+        var g = xform.GridUid;
+        if (g is not { } grid)
+            return;
+        if (!TryComp<MapGridComponent>(grid, out var gridComp))
+            return;
+        var gridBounds = gridComp.LocalAABB;
```

Comment:
Should this be grid bound? If you space the tile it's on suddenly it does nothing.